platform/stm32f4xx/export/platform/spi_bus.hpp

```diff
@@ -532,106 +641,192 @@ void spi_bus< spi_config >::irq_handler()
-constexpr auto spi_bus< spi_config >::pick_rcc()
+template< class spi_i2s_config >
+constexpr auto spi_i2s_bus< spi_i2s_config >::pick_rcc()
 {
     // TODO: comments
-    switch (spi_config::m_dev) {
-    case spi_device::bus_1:
+    switch (spi_i2s_config::m_dev) {
+    case spi_i2s_device::bus_1:
         return RCC_APB2Periph_SPI1;
-    case spi_device::bus_2:
+    case spi_i2s_device::bus_2:
         return RCC_APB1Periph_SPI2;
-    case spi_device::bus_3:
+    case spi_i2s_device::bus_3:
         return RCC_APB1Periph_SPI3;
-    case spi_device::bus_4:
+    case spi_i2s_device::bus_4:
         return RCC_APB2Periph_SPI4;
-    case spi_device::bus_5:
+    case spi_i2s_device::bus_5:
         return RCC_APB2Periph_SPI5;
-    case spi_device::bus_6:
+    case spi_i2s_device::bus_6:
         return RCC_APB2Periph_SPI6;
     default:
         // TODO: clarify
         return static_cast< decltype(RCC_APB2Periph_SPI6) >(-1);
     }
 }
 
-template< class spi_config >
-constexpr auto spi_bus< spi_config >::pick_rcc_fn()
+template< class spi_i2s_config >
+constexpr auto spi_i2s_bus< spi_i2s_config >::pick_rcc_fn()
 {
     // APB1 - SPI3 SPI2
     // APB2 - SPI5 SPI6 SPI1 SPI4
-    switch (spi_config::m_dev) {
-    case spi_device::bus_1:
-    case spi_device::bus_5:
-    case spi_device::bus_4:
-    case spi_device::bus_6:
+    switch (spi_i2s_config::m_dev) {
+    case spi_i2s_device::bus_1:
+    case spi_i2s_device::bus_5:
+    case spi_i2s_device::bus_4:
+    case spi_i2s_device::bus_6:
         return RCC_APB2PeriphClockCmd;
-    case spi_device::bus_2:
-    case spi_device::bus_3:
+    case spi_i2s_device::bus_2:
+    case spi_i2s_device::bus_3:
         return RCC_APB1PeriphClockCmd;
     default:
         // TODO: clarify
         return static_cast< decltype(&RCC_APB2PeriphClockCmd) >(nullptr);
     }
 }
 
-template< class spi_config >
-auto spi_bus< spi_config >::pick_pclk()
+template< class spi_i2s_config >
+void spi_i2s_bus< spi_i2s_config >::init_dma()
+{
+    dma::init_rcc< spi_i2s_config::m_dma_rx_stream >();
+    dma::init_rcc< spi_i2s_config::m_dma_tx_stream >();
+
+    auto handler = [this]() {
+        this->irq_handler();
+    };
+
+    dma::subscribe_irq< spi_i2s_config::m_dma_rx_stream >(handler);
+    dma::subscribe_irq< spi_i2s_config::m_dma_tx_stream >(handler);
+}
+
+template< class spi_i2s_config >
+ecl::err spi_i2s_bus< spi_i2s_config >::i2s_set_audio_frequency(i2s_audio_frequency frequency)
 {
-    RCC_ClocksTypeDef clkcfg;
-    RCC_GetClocksFreq(&clkcfg);
-
-    switch (spi_config::m_dev) {
-    case spi_device::bus_1:
-    case spi_device::bus_5:
-    case spi_device::bus_4:
-    case spi_device::bus_6:
-        return clkcfg.PCLK2_Frequency;
-    case spi_device::bus_2:
-    case spi_device::bus_3:
-        return clkcfg.PCLK1_Frequency;
+    if (spi_i2s_config::m_interface != spi_i2s_interface::i2s) {
+        return ecl::err::notsup;
+    }
+
+    uint32_t tmp = I2S_AudioFreq_8k;
+
+    switch (frequency) {
```

Comment:
Let's develop something to avoid large lookup tables when using enum class and run-time conversion